addons/udes_stock/models/stock_picking.py

```diff
@@ -2753,3 +2753,46 @@ def _get_pick_details(self):
             "submitted_value": self.origin,
         }
         return pick_details
+
+    @api.model
+    def _set_original_picking_on_historical_pickings(self):
+        """
+        Set the Original Picking for historical pickings that do not have an Original Picking set.
+
+        Carries out update as superuser to avoid any permissions issues, as the method is simply
+        setting Original Picking to the correct value, and doesn't take any input variables.
+        """
+        Picking = self.env["stock.picking"]
+
+        pickings_to_update = Picking.sudo().with_context(prefetch_fields=False).search(
+            [("u_original_picking_id", "=", False)], order="id"
+        )
+
+        picking_ids = pickings_to_update.ids
+        picking_count = len(picking_ids)
+
+        if picking_count:
+            _logger.info("Setting Original Picking value on %s pickings", picking_count)
+
+            update_datetime = fields.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S.%f")
+
+            update_query = """
+            UPDATE stock_picking
+            SET
+            write_uid = {user_id},
+            write_date = '{update_datetime}',
+            u_original_picking_id =
+            CASE
+            WHEN backorder_id IS NOT null THEN backorder_id
+            ELSE id
+            END
+            WHERE
+            id IN ({picking_ids})
+            """.format(
+                user_id=SUPERUSER_ID,
+                update_datetime=update_datetime,
+                picking_ids=" ,".join(str(p_id) for p_id in picking_ids),
+            )
+            self.env.cr.execute(update_query)
```

Comment:
We can restructure slightly to follow best practice (not using string formatting for values):

```python
update_query = """
UPDATE stock_picking
SET
write_uid = %(user_id)s,
write_date = now() at time zone 'UTC'
u_original_picking_id =
CASE
WHEN backorder_id IS NOT null THEN backorder_id
ELSE id
END
WHERE
id IN %(picking_ids)s
"""
values = {'user_id': SUPERUSER_ID, 'picking_ids': tuple(picking_ids)} 
self.env.cr.execute(update_query, values)
```

`psycopg2` will convert the tuple of ids into something `IN` can handle.

Reply:
Yep fair point, have now updated the method to be a bit more proper. Although the query has changed somewhat due to backorder fun...!